### ISSUE TITLE
request help: apisix enforces the translation of routed domain names to IP addresses

### ISSUE THREAD

**Issue by 2298814528** [apache/apisix#6296]
### Issue description

I want to use apisix gateway to forward requests to backend **HTTPS service**, but apisix will convert my configured domain name to IP and then forward it. Since the back-end service can only be accessed using domain names and does not support IP access, the forwarding fails. Is there a way to make apisix not automatically convert domain names to IPs?

### Environment

- apisix version (cmd: `apisix version`):2.8
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`): openresty/1.19.9.1
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):3.4.0
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

**Comment by liangliang4ward** [comment]
![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)

can use domain rewrite

**Comment by 2298814528** [comment]
Excuse me, is this a plugin? If yes, please also tell me the name of the plugin, thanks!

**Comment by liangliang4ward** [comment]
> > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > can use domain rewrite
> 
> > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > can use domain rewrite
> 
> Excuse me, is this a plugin? If yes, can you tell me the name of the plugin, thanks!
use proxy-rewirte , add host value to your real host 
"proxy-rewrite": {
            "uri": "/test/home.html",
            "scheme": "http",
            "host": "iresty.com"
           
        }
https://apisix.apache.org/docs/apisix/plugins/proxy-rewrite/

**Comment by 2298814528** [comment]
> > > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > > can use domain rewrite
> > 
> > 
> > > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > > can use domain rewrite
> > 
> > 
> > Excuse me, is this a plugin? If yes, can you tell me the name of the plugin, thanks!
> > use proxy-rewirte plugin , add host value to your real host
> > "proxy-rewrite": {
> > "uri": "/test/home.html",
> > "scheme": "http",
> > "host": "iresty.com"
> > }
> > https://apisix.apache.org/docs/apisix/plugins/proxy-rewrite/

ok thanks, i'll try it now

**Comment by 2298814528** [comment]
> > > > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > > > can use domain rewrite
> > > 
> > > 
> > > > ![image](https://user-images.githubusercontent.com/3895272/153568799-8f3746ca-1e83-451d-8685-23e93a9bf440.png)
> > > > can use domain rewrite
> > > 
> > > 
> > > Excuse me, is this a plugin? If yes, can you tell me the name of the plugin, thanks!
> > > use proxy-rewirte plugin , add host value to your real host
> > > "proxy-rewrite": {
> > > "uri": "/test/home.html",
> > > "scheme": "http",
> > > "host": "iresty.com"
> > > }
> > > https://apisix.apache.org/docs/apisix/plugins/proxy-rewrite/
> 
> ok thanks, i'll try it now

OK, but if there are multiple backend services in my upstream, how should my host be written?
![image](https://user-images.githubusercontent.com/63707437/153573423-eb42aad5-b65b-458d-b74f-e4b8c8d6e7e1.png)

**Comment by tokers** [comment]
@2298814528 Does the problem occur during SSL handshaking since APISIX doesn't send the correct SNI? Or it's caused since APISIX doesn't carry the correct HTTP Host header?

**Comment by liangliang4ward** [comment]
{
  "methods":["GET"],
  "uri": "/test/index.html",
  "plugins":{
      "proxy-rewrite": {
            "host": "iresty.com"      
        }
  }
}

**Comment by tzssangglass** [comment]
> OK, but if there are multiple backend services in my upstream, how should my host be written?

You have to use the same domain name for all these upstream nodes.

**Comment by spacewander** [comment]
> > OK, but if there are multiple backend services in my upstream, how should my host be written?
> 
> You have to use the same domain name for all these upstream nodes.

No. We can use different hosts for different nodes with the `pass_host = node`.
See https://apisix.apache.org/docs/apisix/admin-api#request-body-parameters-3

**Comment by tzssangglass** [comment]
> No. We can use different hosts for different nodes with the `pass_host = node`.
> See [apisix.apache.org/docs/apisix/admin-api#request-body-parameters-3](https://apisix.apache.org/docs/apisix/admin-api#request-body-parameters-3)

I know this. His situation is as follows:

1. the back-end service can only be accessed using domain names and does not support IP access
2. config ip in upstream conf

It seems that in this case he cannot use `pass_host = node`

**Comment by spacewander** [comment]
> the back-end service can only be accessed using domain names and does not support IP access

I guess the upstream is unable to handle direct IP access because they are HTTPS and will check the certificate with the SNI.

It's expected that APISIX will set the SNI according to the host, so using `pass_host = node` should work.

Let's wait for the author's response.

**Comment by liangliang4ward** [comment]
I had the same problem today；

upstream services use web protection，can't direct access use ip.

I use `proxy rewrite plugin` to sovle:
```sql
{
  "uri": "/hello",
  "methods": [
    "POST"
  ],
  "plugins": {
    "proxy-rewrite": {
      "host": "xxx.domain.xxx:85",
      "scheme": "http"
    }
  },
  "upstream": {
    "nodes": [
      {
        "host": "xxx.domain.xxx",
        "port": 85,
        "weight": 100
      }
    ]
  },
  "status": 1
}
```

**Comment by spacewander** [comment]
> I had the same problem today；
> 
> upstream services use web protection，can't direct access use ip.
> 
> I use `proxy rewrite plugin` to solve:
> 
> ```sql
> {
>   "uri": "/hello",
>   "methods": [
>     "POST"
>   ],
>   "plugins": {
>     "proxy-rewrite": {
>       "host": "xxx.domain.xxx:85",
>       "scheme": "http"
>     }
>   },
>   "upstream": {
>     "nodes": [
>       {
>         "host": "xxx.domain.xxx",
>         "port": 85,
>         "weight": 100
>       }
>     ]
>   },
>   "status": 1
> }
> ```

Could you try `pass_host = node` in the upstream? It's expected to modify the host according to the node's host.

**Comment by liangliang4ward** [comment]
> pass_host

我试了一下。
我们有个接口，使用pass_host=node是不行的。有可能是因为，对方防护的时候是判断了host:port的方式（源接口为：xxx.domian.xxx:85）。
但是使用mockbin.org的接口，是可以通过的

**Comment by tzssangglass** [comment]
> 对方防护的时候是判断了host:port的方式

I think you need to investigate the protection rules first and then choose the `pass_host` mode based on the protection rules.

**Comment by 2298814528** [comment]
> I had the same problem today；
> 
> upstream services use web protection，can't direct access use ip.
> 
> I use `proxy rewrite plugin` to solve:
> 
> ```sql
> {
>   "uri": "/hello",
>   "methods": [
>     "POST"
>   ],
>   "plugins": {
>     "proxy-rewrite": {
>       "host": "xxx.domain.xxx:85",
>       "scheme": "http"
>     }
>   },
>   "upstream": {
>     "nodes": [
>       {
>         "host": "xxx.domain.xxx",
>         "port": 85,
>         "weight": 100
>       }
>     ]
>   },
>   "status": 1
> }
> ```

Yes, I have also used this plug-in, which can solve the current problem, but I have a special situation: there are multiple nodes in my upstream nodes (each node is a separate domain name, and node Only supports domain name access), but the host in proxy_rewrite can only write one IP domain name. In this case, apisix can only forward to one node, that is, the value of the host of proxy_rewrite

**Comment by 2298814528** [comment]
> 



> { "methods":["GET"], "uri": "/test/index.html", "plugins":{ "proxy-rewrite": { "host": "iresty.com" } } }

OK, thanks

**Comment by liangliang4ward** [comment]
> > I had the same problem today；
> > upstream services use web protection，can't direct access use ip.
> > I use `proxy rewrite plugin` to solve:
> > ```sql
> > {
> >   "uri": "/hello",
> >   "methods": [
> >     "POST"
> >   ],
> >   "plugins": {
> >     "proxy-rewrite": {
> >       "host": "xxx.domain.xxx:85",
> >       "scheme": "http"
> >     }
> >   },
> >   "upstream": {
> >     "nodes": [
> >       {
> >         "host": "xxx.domain.xxx",
> >         "port": 85,
> >         "weight": 100
> >       }
> >     ]
> >   },
> >   "status": 1
> > }
> > ```
> 
> Yes, I have also used this plug-in, which can solve the current problem, but I have a special situation: there are multiple nodes in my upstream nodes (each node is a separate domain name, and node Only supports domain name access), but the host in proxy_rewrite can only write one IP domain name. In this case, apisix can only forward to one node, that is, the value of the host of proxy_rewrite

You can try this method：
![image](https://user-images.githubusercontent.com/3895272/154393142-c8eecfb8-6cc5-470b-b7e7-d606344cf048.png)

**Comment by 2298814528** [comment]
> > > I had the same problem today；
> > > upstream services use web protection，can't direct access use ip.
> > > I use `proxy rewrite plugin` to solve:
> > > ```sql
> > > {
> > >   "uri": "/hello",
> > >   "methods": [
> > >     "POST"
> > >   ],
> > >   "plugins": {
> > >     "proxy-rewrite": {
> > >       "host": "xxx.domain.xxx:85",
> > >       "scheme": "http"
> > >     }
> > >   },
> > >   "upstream": {
> > >     "nodes": [
> > >       {
> > >         "host": "xxx.domain.xxx",
> > >         "port": 85,
> > >         "weight": 100
> > >       }
> > >     ]
> > >   },
> > >   "status": 1
> > > }
> > > ```
> > 
> > 
> > Yes, I have also used this plug-in, which can solve the current problem, but I have a special situation: there are multiple nodes in my upstream nodes (each node is a separate domain name, and node Only supports domain name access), but the host in proxy_rewrite can only write one IP domain name. In this case, apisix can only forward to one node, that is, the value of the host of proxy_rewrite
> 
> You can try this method： ![image](https://user-images.githubusercontent.com/3895272/154393142-c8eecfb8-6cc5-470b-b7e7-d606344cf048.png)

OK, I'll try it now, thanks!

**Comment by 2298814528** [comment]
> > pass_host
> 
> 我试了一下。 我们有个接口，使用pass_host=node是不行的。有可能是因为，对方防护的时候是判断了host:port的方式（源接口为：xxx.domian.xxx:85）。 但是使用mockbin.org的接口，是可以通过的

对，我这边试了一下也不行
